database.sql

```diff
@@ -309,6 +309,20 @@ CREATE TABLE IF NOT EXISTS `2fa_trusted_browser` (
 	FOREIGN KEY (`uid`) REFERENCES `user` (`uid`) ON UPDATE RESTRICT ON DELETE CASCADE
 ) DEFAULT COLLATE utf8mb4_general_ci COMMENT='Two-factor authentication trusted browsers';
 
+--
+-- TABLE account-user
+--
+CREATE TABLE IF NOT EXISTS `account-user` (
+	`id` int unsigned NOT NULL auto_increment COMMENT 'sequential ID',
```

Comment:
The plan is that this table eventually will be a direct replacement for the contact table. So at one time we will switch over to this table.

Reply:
We need to have a deeper discussion about this. The original contact table was awful because it had only one unique key which was the sequential ID, which meant redundant information for each user having the same contact, and duplicate contacts that you've been trying to remove.

The replacement table shouldn't have a sequential ID unique key, it still won't work. The unique key of the `contact` table should be the `uri-id` on its own, and a separate table should join the `contact` table and the `user` table to record relationships, blocks, etc... which is what we're doing with `user-contact` and should be expanded to include all the user-specific contact settings.

That's it. No sequential ID, no foreign key to the `user` table in the replacement `contact` table. Both of these led us to the present situation, and it's time for them to go.

Reply:
Thing is, we will have to perform some intermediate steps to get to the target point. It's just like the `post-user` table inherited the `id` from the `item` table. This meant that is was possible to transform the data step by step, just changing the source of the `id` but keeping their meaning. By using the same technique, we should be able keeping all the related fields without the need to deprecate all old fields and having to create replacement fields.

My idea for the next development version (2022.12) is to switch the `id` source from the `contact` table to the `account-user` table. So then the `insert` function will at first create an `account-user` entry and will use that `id` for the `contact` table. 

Problem is that (at least) on larger machines there seem to be huge problems with the locking mechanism that we currently use to more or less ensure to don't have duplicates. Several times a day the locking fails and contacts aren't created. And at the same time we still have got problems with duplicates, so there is some need to make changes here.

Reply:
I agree, the change is having only one record for each contact, local or remote. This is why having the `uri-id` as a primary key is a good thing. Can't insert a duplicate if the `uri-id` already is there, locking or not. I don't believe having an intermediary table that looks almost the same as the current table will help us achieve data consistency.

Reply:
Also in the future we will have got user depending fields, so a table with these fields does make sense. In addition to this, we then need a table for the general data and most likely a table per network type. The tables per network type and the general table should have got the `uri-id` as unique key. Then we can create a view that can server as a complete replica of the `contact` table, thus making the switch much easier.

Reply:
If you have a contact table per network type, how do you ensure there aren't any duplicates across different tables? I personally don't like this plan at all, which means that if you go forward with it, you'll be on your own maintaining it. Maybe this isn't a big deal for you, but if you expect any support from other Friendica developers it would be good if you shared your large-scale refactoring plans in advance so we can discuss them.

Reply:
Since the unique key for the network related fields would be the `uri-id`, there can only be a single record per contact and network. But there can be entries for the same contact in several tables, since Friendica and Hubzilla are able to speak both AP and Diaspora. You can have a look at the already existing view `account-user-view`: https://github.com/friendica/friendica/blob/2022.09-rc/static/dbview.config.php#L1000

Of course that view will need some refinements on the way. But this is some groundwork that I would like to use.